File: QSoft.ETW/KernelTrace.cs
using System;
using System.Runtime.InteropServices;

namespace QSoft.ETW;

public sealed class KernelTrace : IDisposable
{
    private const uint ErrorInvalidState = 0xC00000BBu;

    private IntPtr _libraryHandle;
    private readonly StartKernelTraceDelegate? _startKernelTrace;
    private readonly StartHeapTraceDelegate? _startHeapTrace;
    private readonly UpdateHeapTraceDelegate? _updateHeapTrace;
    private readonly CreateMergedTraceFileDelegate? _createMergedTraceFile;

    public KernelTrace(string? libraryPath = null)
    {
        libraryPath ??= "KernelTraceControl.dll";
        _libraryHandle = NativeLibrary.Load(libraryPath);

        _startKernelTrace = LoadFunction<StartKernelTraceDelegate>("StartKernelTrace");
        _startHeapTrace = LoadFunction<StartHeapTraceDelegate>("StartHeapTrace");
        _updateHeapTrace = LoadFunction<UpdateHeapTraceDelegate>("UpdateHeapTrace");
        _createMergedTraceFile = LoadFunction<CreateMergedTraceFileDelegate>("CreateMergedTraceFile");
    }

    public bool Initialize() =>
        _startKernelTrace is not null &&
        _startHeapTrace is not null &&
        _updateHeapTrace is not null &&
        _createMergedTraceFile is not null;

    public uint StartKernelTrace(out nint traceHandle, byte[]? properties, uint cStackTracingEventIds)
    {
        if (_startKernelTrace is null)
        {
            Console.Error.WriteLine("StartKernelTrace function not initialized");
            traceHandle = nint.Zero;
            return ErrorInvalidState;
        }

        return _startKernelTrace(out traceHandle, properties, cStackTracingEventIds);
    }

    public uint StartHeapTrace(string sessionName, uint processIdCount, uint[]? processIds)
    {
        if (_startHeapTrace is null)
        {
            Console.Error.WriteLine("StartHeapTrace function not initialized");
            return ErrorInvalidState;
        }

        return _startHeapTrace(sessionName, processIdCount, processIds);
    }

    public uint UpdateHeapTrace(string sessionName, uint processIdCount, uint[]? processIds)
    {
        if (_updateHeapTrace is null)
        {
            Console.Error.WriteLine("UpdateHeapTrace function not initialized");
            return ErrorInvalidState;
        }

        return _updateHeapTrace(sessionName, processIdCount, processIds);
    }

    public uint CreateMergedTraceFile(string mergedFileName, string[]? traceFileNames, uint traceFileCount, uint extendedDataFlags)
    {
        if (_createMergedTraceFile is null)
        {
            Console.Error.WriteLine("CreateMergedTraceFile function not initialized");
            return ErrorInvalidState;
        }

        return _createMergedTraceFile(mergedFileName, traceFileNames, traceFileCount, extendedDataFlags);
    }

    private TDelegate? LoadFunction<TDelegate>(string functionName)
        where TDelegate : Delegate
    {
        IntPtr functionPtr = NativeLibrary.GetExport(_libraryHandle, functionName);
        if (functionPtr == IntPtr.Zero)
        {
            Console.Error.WriteLine($"GetExport failed for {functionName}");
            return null;
        }

        return Marshal.GetDelegateForFunctionPointer<TDelegate>(functionPtr);
    }

    public void Dispose()
    {
        if (_libraryHandle != IntPtr.Zero)
        {
            NativeLibrary.Free(_libraryHandle);
            _libraryHandle = IntPtr.Zero;
        }
    }

    [UnmanagedFunctionPointer(CallingConvention.Winapi)]
    private delegate uint StartKernelTraceDelegate(out nint traceHandle, byte[]? properties, uint cStackTracingEventIds);

    [UnmanagedFunctionPointer(CallingConvention.Winapi, CharSet = CharSet.Unicode)]
    private delegate uint StartHeapTraceDelegate(
        [MarshalAs(UnmanagedType.LPWStr)] string sessionName,
        uint processIdCount,
        [MarshalAs(UnmanagedType.LPArray, ArraySubType = UnmanagedType.U4)] uint[]? processIds);

    [UnmanagedFunctionPointer(CallingConvention.Winapi, CharSet = CharSet.Unicode)]
    private delegate uint UpdateHeapTraceDelegate(
        [MarshalAs(UnmanagedType.LPWStr)] string sessionName,
        uint processIdCount,
        [MarshalAs(UnmanagedType.LPArray, ArraySubType = UnmanagedType.U4)] uint[]? processIds);

    [UnmanagedFunctionPointer(CallingConvention.Winapi, CharSet = CharSet.Unicode)]
    private delegate uint CreateMergedTraceFileDelegate(
        [MarshalAs(UnmanagedType.LPWStr)] string mergedFileName,
        [MarshalAs(UnmanagedType.LPArray, ArraySubType = UnmanagedType.LPWStr)] string[]? traceFileNames,
        uint traceFileCount,
        uint extendedDataFlags);
}

File: ConsoleApplication_ETW/KernelTrace.cpp
#include "KernelTrace.h"
#include <stdio.h>

KernelTrace::KernelTrace()
    : hLib(NULL),
    pfnStartKernelTrace(NULL),
    pfnStartHeapTrace(NULL),
    pfnUpdateHeapTrace(NULL),
    pfnCreateMergedTraceFile(NULL)
{
    Initialize();
}

KernelTrace::~KernelTrace()
{
    if (hLib != NULL)
    {
        FreeLibrary(hLib);
        hLib = NULL;
    }
}

bool KernelTrace::Initialize()
{
    // 動態載入 KernelTraceControl.dll
    hLib = LoadLibraryW(L"KernelTraceControl.dll");
    if (hLib == NULL)
    {
        wprintf(L"Failed to load KernelTraceControl.dll: %lu\n", GetLastError());
        return false;
    }

    // 載入所有四個 API 函數
    bool success = true;
    success &= LoadFunction("StartKernelTrace", (void*&)pfnStartKernelTrace);
    success &= LoadFunction("StartHeapTrace", (void*&)pfnStartHeapTrace);
    success &= LoadFunction("UpdateHeapTrace", (void*&)pfnUpdateHeapTrace);
    success &= LoadFunction("CreateMergedTraceFile", (void*&)pfnCreateMergedTraceFile);

    if (!success)
    {
        FreeLibrary(hLib);
        hLib = NULL;
        return false;
    }

    return true;
}

bool KernelTrace::LoadFunction(const char* functionName, void*& functionPtr)
{
    functionPtr = GetProcAddress(hLib, functionName);
    if (functionPtr == NULL)
    {
        wprintf(L"GetProcAddress failed for %S: %lu\n", functionName, GetLastError());
        return false;
    }
    return true;
}

ULONG KernelTrace::StartKernelTrace(
    PTRACEHANDLE TraceHandle,
    PEVENT_TRACE_PROPERTIES Properties,
    ULONG cStackTracingEventIds)
{
    if (pfnStartKernelTrace == nullptr)
    {
        wprintf(L"StartKernelTrace function not initialized\n");
        return ERROR_INVALID_STATE;
    }
    return pfnStartKernelTrace(TraceHandle, Properties, cStackTracingEventIds);
}

ULONG KernelTrace::StartHeapTrace(
    const wchar_t* sessionName,
    ULONG processIdCount,
    const ULONG* processIds)
{
    if (pfnStartHeapTrace == NULL)
    {
        wprintf(L"StartHeapTrace function not initialized\n");
        return ERROR_INVALID_STATE;
    }
    return pfnStartHeapTrace(sessionName, processIdCount, processIds);
}

ULONG KernelTrace::UpdateHeapTrace(
    const wchar_t* sessionName,
    ULONG processIdCount,
    const ULONG* processIds)
{
    if (pfnUpdateHeapTrace == NULL)
    {
        wprintf(L"UpdateHeapTrace function not initialized\n");
        return ERROR_INVALID_STATE;
    }
    return pfnUpdateHeapTrace(sessionName, processIdCount, processIds);
}

ULONG KernelTrace::CreateMergedTraceFile(
    const wchar_t* mergedFileName,
    const wchar_t** traceFileNames,
    ULONG traceFileCount,
    DWORD extendedDataFlags)
{
    if (pfnCreateMergedTraceFile == NULL)
    {
        wprintf(L"CreateMergedTraceFile function not initialized\n");
        return ERROR_INVALID_STATE;
    }
    return pfnCreateMergedTraceFile(mergedFileName, traceFileNames, traceFileCount, extendedDataFlags);
}